SAI-5089 : Slow node detector to return quicker partial results from QA node

patsonluk opened the pull request:
## Description
A module in QA (coordinator) node such that once a slow node is detected, it should avoid waiting for too long on such node and start returning partial result with shorter wait time. For complete design and discussions, please refer to this Google [Doc](https://docs.google.com/document/d/169VJ3DN2TsEZv2UduGuwCXCt_lQqFnfYNHsqi6kXPbw/edit?usp=sharing)

## Solution
### HttpShardHandler/TimeLimitingHttpShardHandler
Changes to existing `HttpShardHandler` to notify for each `ShardRequest` submission and completion to a shard url. This could be quite similar to what existing `TrackingShardHandlerFactory` provides, however, it only notifies on submission and lacks several critical info for our purpose (the relevant `Future`, shard urls etc) therefore I decided to add the "hooks" in `HttpShardHandler` and create a new `TimeLimitingHttpShardHandler` instead. Such handler would register a list of `ShardRequestActor`s (more in next section) and notify them on each `ShardRequest` submission and completion to all the shards.

`TimeLimitingHttpShardHandler` should only be used as `shardHandlerFactory` for `SearchHandler`, other places should use the default `HttpShardHandler` as `shardHandlerFactory`. This is achieved by setting `shardHandlerFactory` in the `solrconfig.xml`  as described in [here](https://solr.apache.org/guide/solr/latest/configuration-guide/requesthandlers-searchcomponents.html#search-handlers) while leaving the default settings in `solr.xml`

### SlowNodeDetector
A stateful detector to determine and maintain a list of "slow nodes" based on previous request latency stats. For details, please refer to the [design doc](https://docs.google.com/document/d/169VJ3DN2TsEZv2UduGuwCXCt_lQqFnfYNHsqi6kXPbw/edit?usp=sharing).

There's one instance per `TimeLimitingHttpShardHandlerFactory` instance. And usually the Solr process would only have one `TimeLimitingHttpShardHandlerFactory` instance. 

### ShardRequestActor
An actor tied to a particular `ShardRequest` instance (a single `ShardRequest` can be submitted to many shards, the class itself is a bit confusing imho...) which gets notified on submission and completion of such `ShardRequest` instance to each shard. There are current 2 actors and are initialized by `TimeLimitingHttpShardHandler` (hard-coded for now)
1. `NodeStatsCollector` - An actor that keep tracks of the response latency and report the stats to `SlowNodeDector`. 
2. `SlowNodeTimeoutActor` - Keeps track of the `ShardRequest` submissions that are pending response, and issue a delayed timeout if all of the outstanding ones are from the "slow node" list. Refer to the Google doc for details. This is only applied to request with `shards.tolerant=true`

### Metrics
Exposed as `solr.node` metrics in solr/admin/metrics:
1. `QUERY.httpShardHandler.cancelledSlowNodeRequests` - cumulative count of requests to shard cancelled due to slow node
2. `QUERY.httpShardHandler.cancelledDryRunSlowNodeRequests` - cumulative count of requests to shard would have cancelled due to slow node but did not, due to dry-run mode
3. `QUERY.httpShardHandler.slowNodeCount` - a snapshot of slow node count
4. `QUERY.httpShardHandler.slowNodes` - a snapshot of the list of slow Solr nodes (string array)

Take note that 1., 2. and 3. are also available in solr/metrics as `cancelled_slow_node_requests`, `cancelled_dry_run_slow_node_requests` and `slow_node_count` 


## Test
Tests conducted on playpen by deploying this [branch](https://github.com/cowpaths/fs-solr/tree/patsonluk/SAI-5089/playpen-test)  to playpen and configured solr-c92-1 to have 30 artificial delays on `/select` query. Then solr-bench drive query traffic to 100 collections concurrently (each with 1 thread)
### Functionality test
1. Verified `shards.tolerant` behavior, 
   1. `shards.tolerant=true`, qa node latency at 5 sec and solr-c92-1 latency at 30 sec
   3. With absence of `shards.tolerant`, both qa node and solr-c92-1 with latency at 30 sec. As without `shards.tolerant`, we will fall back to existing behavior (no timeout interrupt). However, slow node c92-1 was still detected
2. Confirmed logging 
4. Confirmed the metrics values are correct (both solr/metrics and solr/admin/metrics) 
5. dry-run mode - Log prints `Dry-run mode: would have cancelled 2 pending requests...`, while both QA and Data node wait for full 30 secs
6. Confirmed other config params working as expected

** Re-testing due to many changes, found a bug in `SearchHandler` that shardHandler was missing init metrics **

### Benchmarking
Minor overhead on QA node when there are slow nodes detected and short circuiting:
1. More threads when there are timeouts on slow nodes (1 per top level slow ShardRequest). However, given the short timeout (5 sec) and pretty low query rate on our prod  1-4 per sec per qa node (once a week spike to 50+, but rare), the extra threads should not have any impact
2. CPU/GC impact very minor even in extreme case (everything on the slow node timeout, all response from such node are slow etc)
3. When there's no slow node. There's no noticeable difference in anything.

Details below:
#### Scenario 1: solr-c92-1 as slow node with 30 secs artificial delay
100 collections (8 cores each) on playpen c92, then issue query via solrbench on all 100 collections concurrently with 1 thread each, The deployed hashes/test timeframes are:
- 16:23 - 18:03 : Existing fs-solr hash (`a1c20db1c03f76365c71db7c8fc186b6a73f76e3 `) w/o changes proposed in this PR
- 18:11 - 21:04: Test [branch](https://github.com/cowpaths/fs-solr/tree/patsonluk/SAI-5089/playpen-test) dry-run = true
- 21:37 - 23:24: Test [branch](https://github.com/cowpaths/fs-solr/tree/patsonluk/SAI-5089/playpen-test) dry-run = false

![image](https://github.com/user-attachments/assets/61343657-d9b8-4c15-8f92-15af620f7fb4)
^^ Latencies are correct as the first 2 periods, the both qa node and data node c92-1 show 30 sec latency, while in the last period, qa node is down to 5 sec (timed out on partial result) and data node c92-1 is 30 sec latency

![image](https://github.com/user-attachments/assets/1034e0ae-dbc7-4cd3-b704-613eb3ed87ed)
^^ Thread count, the non dry-run mode uses around 100 more threads, this is expected as there's one timeout thread per  shard request and we have around 100 concurrent requests (1 per collection, 100 collection). 

![image](https://github.com/user-attachments/assets/2f7c7ad3-3651-4635-b054-40650214c4b6)
^^ No full GC, relatively higher young gen/mixed GC percentage, but it's still very low

**Conclusion: Expected extra resource usage with the change. However, it's minimal even with the test rate of 20 req/sec on qa node compared to prod which usually in the range of 1-4 (once a week spike to 50+, but rare)**

#### Scenario 2: no slow node
Since solr-c92-1 has some unexplained slowness while carrying out the tests, moved the tests to solrperf environment.

100 collections (8 cores each) on solrperf cluster (2 data/1 qa/1os), then issue query via solrperf/solrbench on all 100 collections concurrently with 1 thread each, The deployed hashes/test timeframes are:

- 20:06 - 21:06 : Existing fs-solr hash (`a1c20db1c03f76365c71db7c8fc186b6a73f76e3 `) w/o changes proposed in this PR
- 18:02 - 19:02 : [branch](https://github.com/cowpaths/fs-solr/tree/patsonluk/SAI-5089/9-7-test) point at fullstory-solr [branch](https://github.com/cowpaths/fullstory-solr/tree/patsonluk/SAI-5089/9_7-node-heuristics) with changes from this PR but based on [fs/branch_9_7](https://github.com/cowpaths/fullstory-solr/tree/fs/branch_9_7) in fullstory-solr dry-run = false :

Take note that we did not test on dry-run = true as it should just be the same as dry-run = false as there are no slow nodes

![image](https://github.com/user-attachments/assets/d9a90de2-63ce-41ad-9510-57fa653dcf88)
^^ latencies very similar

Not posing other metrics for brevity as they are almost identical

**Conclusion: No noticeable difference when there're no slow nodes**

patsonluk committed:
Initial commit of TimeLimitedHttpShardHandler

patsonluk committed:
Reduced changes on HttpShardHandler

patsonluk committed:
Cancel all pending futures

patsonluk committed:
Rewrote the TimeLimitedHttpShardHandler to handle different shard requests

Added unit test case skeleton

patsonluk committed:
Fixed Timer usage
Improved unit test cases

patsonluk committed:
Added node heuristic to determine "slow nodes" and to timeout on those requests

patsonluk committed:
Code cleanup

patsonluk committed:
Test cases and further cleanup of the logic

patsonluk committed:
Use MDC threadpool instead of timer

patsonluk committed:
Unit test case for slow node detector.
Renamed some variable to be more descriptive

patsonluk committed:
Unit test case for slow node detector.
Renamed some variable to be more descriptive
Code cleanup

patsonluk committed:
1. Renamed config param from timeout to slowNodeTimeout to avoid confusion since it will be used alongside with other existing shardHandlerFactory params
2. Added node metrics QUERY.cancelledSlowNodeRequests and adjusted unit test cases

patsonluk committed:
Fixed implementation so it works for shard with multiple urls as well (> 1 replication factor)

patsonluk committed:
Unit test case for > 1 replication factor

patsonluk committed:
Push the node specific logic to the listener. Make ShardRequestListener an interface

patsonluk committed:
Renamed some classes for consistency with high level discussions

patsonluk committed:
Fixed timeout handling for shard request with both slow and fast url.

Added unit test case

patsonluk committed:
Implement all the configurable values

patsonluk committed:
Added slow node list metrics.

Expose metrics to fullstory prometheus servlet as well

patsonluk committed:
Fixed solr metrics init

patsonluk committed:
Refactor shard request actor to separate the stats collection and timeout duty \o/

Only timeout if shards.tolerant=true

patsonluk committed:
Renamed class for consistencies

Changed DEFAULT_SLOW_NODE_TTL from 60 sec to 60 mins

patsonluk committed:
Fixed javadoc and ./gradlew tidy

patsonluk committed:
./gradlew tidy

patsonluk committed:
Fixed log messages

patsonluk committed:
Fixed log messages

patsonluk committed:
Added log.isInfoEnabled() checks

patsonluk committed:
./gradlew tidy

patsonluk committed:
Added log.isInfoEnabled() checks

patsonluk committed:
Removed the unused javadoc

patsonluk committed:
Removed ussage of com.google.common.collect.MapMaker

patsonluk committed:
./gradlew tidy

patsonluk committed:
shards.tolerant test case
and fix possible NPE

patsonluk committed:
Only call timeoutCallback if dryRun is false, since in dryRun mode we do not time out the actual request

patsonluk committed:
Changed test case solr.xml config
./gradlew tidy

patsonluk committed:
Added missing javadoc
Added new metric counter cancelledDryRunSlowNodeRequests and corresponding test case
Renamed test case for consistency

patsonluk committed:
minor change to test case
./gradlew tidy

patsonluk commented:
Tagging everyone in the team for review as I think this is relevant to us all 😊 

There's still one unit testing issue on Github Actions (it cannot load my test solr.xml config there, but works fine locally), still debugging that one but it should not affect code review

patsonluk committed:
Undo unnecessary change

patsonluk committed:
Fixed javadoc

patsonluk committed:
Fixed test case solr.xml lookup

hiteshk25 commented on solr/core/src/java/org/apache/solr/handler/component/SlowNodeDetector.java:
```diff
@@ -0,0 +1,249 @@
+        break;
+      }
+
+      previousLatency = current.latency;
```
Have you consider `previousLatency` to max latency? I feel like if we see half of the max latency then there is issue.

patsonluk replied:
If i understand correctly, are you suggesting if we find any latencies than is < 0.5 * max latencies, then flag it (flag the max?) 

Would you mind to give a solid example so I understand your proposal better? 😊 

From my investigation so far, having latencies half of the max latency is pretty common and normal. For example in this [ticket](https://fullstory.atlassian.net/browse/VAL-9270), the latencies of top 8 nodes are
![image](https://github.com/user-attachments/assets/2b67f123-28b7-4dce-8f77-3f24716003f0)

while the median is 15sec for all 256 nodes. The above case is NOT a slow node case.

Whereas in a slow node case, we usually observe at least *2 latencies on the slowest node(s) compared to other nodes.  For example in the below case, the top node is 300+ sec, while the 2nd node is 45 sec

![image](https://github.com/user-attachments/assets/58a6472d-118f-4b59-931d-1288f127c55f)

Therefore this PR identifies the "sudden drop" of latencies caused by slow node(s) by iterating the sorted numbers.

hiteshk25 replied:
Have many shards/node you want to consider in request? will that take effect in this case?

patsonluk replied:
The detection logic and the early timeout logic have different criteria right now

For detection, currently `DEFAULT_MIN_CORE_PER_REQUEST = 512` , which means a request has to at least have 512 cores in order to trigger slow node detection (avoid single core having unbalanced shard that trigger slow node, this is not bullet proof, might need to verify once we do dry-run). `DEFAULT_MAX_SLOW_NODE_PERCENTAGE = 10` controls how many requests should be iterated to detect slow node. This is to avoid false detection when a lot of nodes seem to be slow as other fast nodes were fast simply due to cache/no docs/other optimizations etc. 

For early timeout, there's no shards/nodes limit explicitly. However, there should be at least one "fast" response before the interrupt logic be triggered, so it has to be > 1  (which is true to all orgs anyway afaik? 😁 )

hiteshk25 commented on solr/core/src/java/org/apache/solr/handler/component/TimeLimitingHttpShardHandler.java:
```diff
@@ -0,0 +1,328 @@
+            ExecutorUtil.newMDCAwareSingleThreadExecutor(
+                new SolrNamedThreadFactory("SlowNodeTimeout"));
+        timeoutTask =
+            executorService.submit(
```
This code I'm not following. Should be part of ` onRequestSubmitted` function, and then it should watch response from slow nodes? Maybe we need to chat on this

patsonluk replied:
This code uses `onRequestCompleted` to achieve a more "adaptive" wait time. 

For example, if we have a timeout set as 5 secs:
1. For a really cheap query. We will give slow node 5 secs to finish. Since all the other nodes complete quickly (at almost no time at all), and then this will get triggered and wait for 5 secs
2. For an expensive query. Say all the other nodes took 30 secs to finish, then we will wait for additional 5 sec for the slow node, hence we will give slow node 35 secs to finish.

If we start the timer at `onRequestSubmitted`, we would not be able to tell how expensive the query will be, hence we cannot assign a wait time dynamically. We could risk timing out the request on a node just because the query itself is expensive and not because of such node being slow.

hiteshk25 commented:
@patsonluk Thanks. Added couple of comments. We can more on Monday for it.

patsonluk committed:
Added extra info messages

magibney commented on solr/core/src/java/org/apache/solr/handler/component/TimeLimitingHttpShardHandler.java:
```diff
@@ -0,0 +1,328 @@
+              actors.add(new NodeStatsCollector(slowNodeDetector));
+              if (shardRequest.params != null
+                  && "true"
+                      .equalsIgnoreCase(shardRequest.params.get(ShardParams.SHARDS_TOLERANT))) {
```
The idiomatic way to do this is:
```java
  ShardParams.getShardsTolerantAsBool(req.getParams())
```

magibney commented on solr/core/src/java/org/apache/solr/handler/component/TimeLimitingHttpShardHandler.java:
```diff
@@ -0,0 +1,328 @@
+        actorsByShardRequest.computeIfAbsent(
+            shardRequest,
+            k -> {
+              List<ShardRequestActor> actors = new ArrayList<>();
```
We know we can size this ArrayList to 2 here. Kind of depends on how big we expect the map to get, whether it's worth it...

patsonluk replied:
I was about to just set it to 2, however, after doing so, the constant gives me a 😐 feeling - especially now such constant will need to be adjusted if more actors are added (super minor)

The map should be quite reasonable in size, it's one shardRequest PER high level request in most cases. Perhaps okay to keep it simple as is for now? 😊

magibney replied:
> gives me a 😐 feeling

Fair 😁 

> map should be quite reasonable in size

👍 this is sufficient to reassure me.

magibney commented on solr/core/src/java/org/apache/solr/handler/component/TimeLimitingHttpShardHandler.java:
```diff
@@ -0,0 +1,328 @@
+              }
+              return actors;
+            });
+    shardRequestActors.forEach(actor -> actor.onRequestSubmitted(shardUrls, future));
```
Could this throw ConcurrentModificationException, iterating over `shardRequestActors` while `ShardRequestTrackingCallback.onComplete()` might be removing actors? Same underlying issue could also manifest as a race condition where actors get removed prematurely, if their `pendingCount` is decremented to 0 while another thread attempts to increase it here.

Perhaps there's a reason why this is not a concern, practically? Worth considering though.

I _think_ this could be an issue if there was some delay in submitting requests, and responses started coming in while subsequent requests are still being submitted.

magibney replied:
Would it be possible to (and would there be any benefit to) leverage `ShardRequest.actualShards.length`, to place a floor on the number of responses to expect? ... could be a failsafe to prevent registering stats for requests that are marked "complete" prematurely?

patsonluk replied:
> Could this throw ConcurrentModificationException, iterating over `shardRequestActors` while `ShardRequestTrackingCallback.onComplete()` might be removing actors? Same underlying issue could also manifest as a race condition where actors get removed prematurely, if their `pendingCount` is decremented to 0 while another thread attempts to increase it here.
> 
> Perhaps there's a reason why this is not a concern, practically? Worth considering though.
> 
> I _think_ this could be an issue if there was some delay in submitting requests, and responses started coming in while subsequent requests are still being submitted.

I think this exposes certain design flaw with my code. 

My original idea is - let's give each `ShardRequestActor` the freedom to tell the caller whenever it wants to stop listening to further shard request event, and when none of the shard request actors are interested in a particular shard request anymore, purge such shard request key from  `actorsByShardRequest`. 

However, this does not work too well as the scope is not well defined. For example if there are total 4 cores, somehow 3 cores are submitted and finished before the last core requests was submitted, then what exactly should the behavior be? After finishing the first 3 cores, the actor probably returned `true` (remove me), but then on 4th request, it will be re-added and re-notify (new entry in `actorsByShardRequest`)

The "re-add" behavior is not wrong imho, but it highlights that the `boolean` returned by `ShardRequestActor#onRequestCompleted` is not well defined. So my proposal - let's remove such ability from actor 😁 (and it's not useful for our case anyway and adds risk to `ConcurrentModificationException` as pointed out by you)

Perhaps, `TimeLimitingHttpShardHandler` will keep track of outstanding requests count and notify actors with a `close`/`flush` function. Especially that we can probably tell if the caller has finished with a particular `ShardRequest` (see next reply section)

> Would it be possible to (and would there be any benefit to) leverage ShardRequest.actualShards.length, to place a floor on the number of responses to expect? ... could be a failsafe to prevent registering stats for requests that are marked "complete" prematurely?

Yes definitely, the major advantage is that we wouldn't have to rely on tracking in-flight requests and can simply count responses to see if everything has completed (hence solving the 3 complete before last submitted scenario above!) However, I was reluctant as:
1. The guarantees of `actualShards` seemed weak to me, the execution logic within `SearchHandler` does not iterate on such list, it is the "caller" which "could" iterate on such list when calling `submit`
2. So i was under the impression that the restriction could be very loose, like one might not need to really set `actualShards` and just call submit with the 2nd `shard` param filled

It turns out that my understanding on 2. is not accurate. `SearchHandler` does enforce `actualShards.length` should be same as `responses.length` at `take` [here](https://github.com/cowpaths/fullstory-solr/blob/fs/branch_9_7/solr/core/src/java/org/apache/solr/handler/component/HttpShardHandler.java#L230) . It's a bit weird, but I take it as a guarantee 😁

magibney replied:
> My original idea is - let's give each ShardRequestActor the freedom to tell the caller whenever it wants to stop listening to further shard request event, and when none of the shard request actors are interested in a particular shard request anymore, purge such shard request key from actorsByShardRequest.

This is intriguing, and I could imagine it being useful; but additional description/documentation/comments/examples would really help here. If we don't have a clearly defined use case in mind, it's better to avoid the additional complexity?

Re: `ShardRequest.actualShards.length`, I was not thinking to _replace_ tracking in-flight requests (I think there's little harm in tracking in-flight requests), but rather just to use it as a failsafe to ignore cases with fewer responses than expected. So basically track in-flight requests as usual, but last step before registering stats would be: "do we have at least `actualShards.length` responses? Yes -> register-stats, No -> ignore". This would protect against the "race condition/re-add" case, as well as any other case that might result in fewer requests than expected (since all our decisions/heuristics are based on the assumption that we have a full complement of responses).

patsonluk replied:
I believe this can be marked as resolve (avoiding ConcurrentModificationException) by changes introduced in https://github.com/cowpaths/fullstory-solr/pull/255? please lemme know

magibney replied:
Agreed, looks good.

magibney commented on solr/core/src/java/org/apache/solr/handler/component/TimeLimitingHttpShardHandler.java:
```diff
@@ -0,0 +1,328 @@
+class Util {
+  static String getNode(String urlString) {
+    try {
+      URL url = new URL(urlString);
```
Prefer URI over URL. I'm pretty sure this URL ctor (and java.net.URL ctors generally) are deprecated as of jdk20. In fact I'm surprised we don't get a deprecation warning that fails the build.

magibney commented on solr/core/src/java/org/apache/solr/handler/component/TimeLimitingHttpShardHandler.java:
```diff
@@ -0,0 +1,328 @@
+
+    private void onComplete(long elapsedTime, String selectedShardUrl) {
+      try {
+        actorsByShardRequest.compute(
```
Is there any reason that we can't just include a reference to the `actors` directly in the callback, to avoid the second lookup in `onComplete()`?

Then in `onComplete()` we'd only need locking access to the map when we know we're ready to remove.

magibney replied:
Hmm I think if we do ^ (avoiding a lookup in the callback), then we might be able to simply use `WeakHashMap` (or probably even just `HashMap` 🤷 ) for `actorsByShardRequest`? and _never_ explicitly remove entries.

This would assume:
1. transient lifecycle for ShardHandler (i.e. they don't hang around long-term)
2. reasonable number of ShardRequest per top-level request

I _think_ these assumptions are valid ...

patsonluk replied:
I have replaced the `actorsByShardRequest` with `activeShardRequests` and use a concurrent map since I agree with the transient nature of the ShardHandler :)

It is also true that there should only be one active ShardRequest at any given time. (deduced from code implementation in HttpShardRequest 😁 ) I do think the Concurrent Map enables easy sync control and is a bit more bullet proof? (in case if HttpShardHandler does support concurrent request in the future...)

magibney commented on solr/core/src/java/org/apache/solr/handler/component/TimeLimitingHttpShardHandler.java:
```diff
@@ -0,0 +1,328 @@
+class NodeStatsCollector implements ShardRequestActor {
+  private static final Logger log = LoggerFactory.getLogger(MethodHandles.lookup().lookupClass());
+  private final SlowNodeDetector detector;
+  private final AtomicInteger pendingCount = new AtomicInteger(0);
```
Is there a reason that we track `pendingCount` at the `ShardRequestActor` level, as opposed to augmenting the ArrayList to carry `pendingCount` state? If that's possible, I think doing so would also address the ConcurrentModificationException concern on ArrayList. (I think there'd still be a race condition, but basically benign)

patsonluk replied:
Yep. Removed the pendingCount from NodeStatsCollector in the latest implementation, the higher level logic now determines whenever `close` should be invoked on the actor, hence flushing the data.

magibney reviewed (commented):
Still going through this, but sharing my comments/questions so far

magibney commented on solr/core/src/java/org/apache/solr/handler/component/TimeLimitingHttpShardHandler.java:
```diff
@@ -0,0 +1,328 @@
+      try {
+        executorService =
+            ExecutorUtil.newMDCAwareSingleThreadExecutor(
+                new SolrNamedThreadFactory("SlowNodeTimeout"));
```
Creating and shutting down a new thread pool in the local scope of this method seems not ideal. Can we have the thread pool live in the ShardHandlerFactory?

patsonluk replied:
For sure! Moved it to `SlowNodeTimeoutActor` as a field. Since `close` is introduced to `ShardRequestActor`, we can invoke `ExecutorService#shutdown` in `SlowNodeTimeoutActor#close` matching the lifecycle 🙌🏼

magibney replied:
This does seem somewhat cleaner; but doesn't fully address my concern about the fact that we're basically spinning up a thread pool (and at least one thread) _per top-level request_, which runs a bit against the spirit of a thread pool.

If we're configuring the ShardHandlerFactory at the SearchHandler level (which I think makes sense), we'd be unable to share threads/thread pool across all collections on the node, which is unfortunate. But still, having the thread pool live in the ShardHandlerFactory (even if one per collection) would allow threads to be reused across multiple top-level requests. I think this would make sense?

patsonluk replied:
As discussed, move it further up to `TimeLimitingHttpShardHandlerFactory` now. Take note that:
1. In the current `SearchHandler`, per core on `inform`, if SearchHandler specific shardHandlerFactory is defined, then it invokes `shardHandlerFactory = core.createInitInstance(shfInfo, ShardHandlerFactory.class, null, null);`, which means each core will have its own instance of `TimeLimitingHttpShardHandlerFactory` (ie threadpool), vs the coreContainer one, which usually is one instance per Solr process.
2. Fortunately, for QA node, there's only one core loaded - the synthetic core, therefore, we will essentially only have one thread pool as desired. For data node, many cores are loaded, but the shardHandlerFactory should not be used as data node should not be distributing requests to other data node.

magibney commented on solr/core/src/java/org/apache/solr/handler/component/TimeLimitingHttpShardHandler.java:
```diff
@@ -0,0 +1,328 @@
+                  } catch (InterruptedException e) {
+                    // ok. The responses came back before timeout
+                    return;
+                  }
```
Might prefer a different construct than interrupt here? Perhaps:
```java
CountDownLatch allComplete = new CountDownLatch(1);
...
  if (allComplete.await(timeout, TimeUnit.MILLISECONDS) {
    // ok. The responses came back before timeout
    return;
  }
```
May only need the CountDownLatch then, not need to the Future reference? If we _do_ keep the Future reference, we should call `future.get()`, to allow logging any exceptions thrown; if we _don't_ keep the Future reference, we should ensure that we catch and log any exceptions _within_ the timeoutTask lambda.

Practically, this also fixes a race condition: If a request completes after the timeout task has cleared `pendingFutures`, it will see `pendingFutures` as empty and attempt to cancel timeout task -- potentially interrupting it before it completes its work. `CountDownLatch` solution (or similar) doesn't have this issue.

patsonluk replied:
@magibney changes [here](https://github.com/cowpaths/fullstory-solr/pull/246/commits/3d3f40e87eebb0271d60376c027e1115c1c01a47) to replace Future reference with `CountDownLatch`.

There're also some extra changes to prevent certain race conditions.

magibney commented on solr/core/src/java/org/apache/solr/handler/component/TimeLimitingHttpShardHandler.java:
```diff
@@ -0,0 +1,328 @@
+                            timeout);
+                      }
+                      removingFutures.forEach(f -> f.cancel(true));
+                      if (log.isInfoEnabled()) {
+                        log.info("{} Pending requests cancelled", removingFutures.size());
+                      }
```
do we really need to log at INFO level before _and_ after, essentially the same message?

patsonluk replied:
Good point! Consolidated them into a single message after they are cancelled

magibney commented on solr/core/src/java/org/apache/solr/handler/component/TimeLimitingHttpShardHandler.java:
```diff
@@ -0,0 +1,328 @@
+                    // ok. The responses came back before timeout
+                    return;
+                  }
+                  Set<Future<?>> removingFutures = new HashSet<>(pendingFutures);
```
curious: is there a reason to make a copy here? It will still iterate over `pendingFutures`, just inside the `HashSet` ctor instead of inline ...

patsonluk replied:
I really don't remember the exact reason 😓 . Perhaps i was trying to minimize holding off `pendingFutures`(ie smaller `synchronized` block) or if there are new requests submitted, I didn't want to cancel them(?), 

Since all the ops here should be fast and existing code has several race-conditions, l propose:
1. Use pendingFutures directly here. Wrap it with `synchronized` block
2. Mark `onRequestSubmitted`, ~`flush`~ `close` as `synchronized` too

What do you think? :)

patsonluk replied:
The above changes were incorporated into this [commit](https://github.com/cowpaths/fullstory-solr/pull/246/commits/3d3f40e87eebb0271d60376c027e1115c1c01a47)

Take note that `flush` is renamed to `close` as once `close` is called, there should be no additional calls to such actor (this avoids quite a few race condition headaches).

Now there are still 2 scenarios I want to highlight:
1. After the countdown is triggered (ie some request completions and all that remains are slow nodes, but before the timeout is reached), there could be new request being submitted (unlikely but 🤷🏼 ), those WILL be subjected to the cancellation as well since it's all the same `pendingFutures`. I think this behavior is fine
2. Current `activeShardRequests.compute` should be a safeguard to prevent additional method invocations on the actors once `close` is invoked.

magibney reviewed (commented):
Left a few more comments/questions; still haven't looked at the actual detection logic/heuristics ...

patsonluk committed:
SAI-5089 : Track active shardRequest in a higher level (instead of actor) (#255)

* Use  ShardParams.getShardsTolerantAsBool

* Determine whether all requests for a shard request are completed on higher level

* Track submitted vs completed requests in ShardRequestTracker instead of relying on actualShards field in HttpShardHandler

* Updated javadoc

patsonluk committed:
Use URI instead of URL

patsonluk committed:
Use cancelCountDownLatch instead of timeoutTask

Fixed some race condition.

Rename `flush` to `close` for ShardRequestActor and defines expected behaviors

patsonluk committed:
Move `ExecutorService` to `SlowNodeTimeoutActor` with better lifecycle management

patsonluk committed:
Consolidate info message

patsonluk committed:
./gradlew tidy

patsonluk committed:
Removed map lookup from ShardRequestTrackingCallback#onComplete, instead pass the tracker to ctor and use it directly

Since we are not doing strict synchronization via `compute` anymore. Relax the `TimeLimitingHttpShardHandler#onRequestSubmit` part to use computeIfAbsent and leave the `onRequestSubmitted` call outside to minimize locking

magibney commented on solr/core/src/java/org/apache/solr/handler/component/TimeLimitingHttpShardHandler.java:
```diff
@@ -67,27 +67,25 @@ protected ShardRequestCallback onRequestSubmit(
+              return new ShardRequestTracker(actors);
+            });
+
+    tracker.actors.forEach(actor -> actor.onRequestSubmitted(shardUrls, future));
+    tracker.outstandingRequestCount.incrementAndGet();
```
may as well put `incrementAndGet` _before_ `actors.forEach(...)`

magibney commented on solr/core/src/java/org/apache/solr/handler/component/TimeLimitingHttpShardHandler.java:
```diff
@@ -0,0 +1,356 @@
+            actor -> actor.onRequestCompleted(selectedShardUrl, shardUrls, future, elapsedTime));
+        int outstandingRequestCount = tracker.outstandingRequestCount.decrementAndGet();
+        if (isLastResponse(isException, outstandingRequestCount)) {
+          tracker.actors.forEach(ShardRequestActor::close);
+
+          // there could be some race condition here but benign
+          // (other thread triggers onRequestSubmit on actors which are already closed etc)
+          activeShardRequests.remove(shardRequest);
```
I think the race condition on `activeShardRequests.remove(shardRequest)` is benign; but there's also a possibility here that `tracker.actors.forEach(ShardRequestActor::close)` will be called multiple times for the same tracker/actors.

patsonluk replied:
As discussed in Slack, we will:
1. Keep the current flow - `computeIfAbsent` in `onRequestSubmit`, no extra synchronization in `onComplete`
2. Relax the requirement in `ShardRequestActor` javadoc, that the implementation should prepare that more method invocations could arrive after `close` is invoked (though unlikely), and they should take care to NOT throw exception/cause issues, the actual behavior is up to the implementation. This is to avoid complications to handle various race-condition due to the previous `ShardRequestActor` requirements, which were stricter

patsonluk committed:
Moved the executorService from Actor to TimeLimitingHttpShardHandlerFactory

patsonluk committed:
javadoc. Fixed incorrect shutdown

patsonluk committed:
avoid NPE if somehow init failed

patsonluk committed:
Minor reordering of statements

patsonluk committed:
Relax the behavior requirement of  ShardRequestActor#close a little bit to avoid overly complicated race-condition handling and unnecessary blocking/map lookup

patsonluk committed:
./gradlew tidy

hiteshk25 reviewed (approved):
Thanks @patsonluk.  LGTM ( let's wait for all Michael's feedback )

patsonluk committed:
Fixed missing metrics init on shardHandlerFactory in SearchHandler

patsonluk committed:
Fixed pendingFutures log message on timeout

patsonluk committed:
./gradlew tidy

magibney commented on solr/core/src/java/org/apache/solr/handler/component/TimeLimitingHttpShardHandlerFactory.java:
```diff
@@ -0,0 +1,180 @@
+  @Override
+  public void close() {
+    super.close();
+    if (executorService != null) {
+      executorService.shutdown();
+    }
```
Prefer `ExecutorUtil.shutdownNowAndAwaitTermination(executorService)` here?

I'm concerned that otherwise this might artificially cause the JVM to block until any existing timeout tasks complete (up to 60s).

patsonluk replied:
Definitely! I always forget about the ExecutorUtil from Solr 😓 , should have used `shutdownNow` regardless!

magibney commented on solr/core/src/java/org/apache/solr/handler/component/TimeLimitingHttpShardHandler.java:
```diff
@@ -0,0 +1,371 @@
+    }
+
+    private boolean isLastResponse(boolean isException, int outstandingRequestCount) {
+      if (!shardsTolerant && isException) {
+        return true;
+      }
```
I don't understand why we return `isLastResponse()==true` for this case?

Wouldn't this basically guarantee that `close()` would be called once per request that completes with an exception? Is this the goal? I think it should be ok, it's a just a little weird, and might benefit from a comment in the code to explain the purpose?

patsonluk replied:
Great question! I think your are right!

The reason I added this is based on this [code](https://github.com/cowpaths/fullstory-solr/blob/fs/branch_9x/solr/core/src/java/org/apache/solr/handler/component/HttpShardHandler.java#L223-L224), which no further response will be added to `rsp.getShardRequest().responses.add(rsp);` (super weird structure 😓 )

However, now I looked at it again. `ShardRequestCallback#onComplete` is tied to the actual `Future` at [here](https://github.com/cowpaths/fullstory-solr/blob/patsonluk/SAI-5089/node-heuristics/solr/core/src/java/org/apache/solr/handler/component/HttpShardHandler.java#L165-L188), which does not care about how responses are added.

I will remove the condition above and ensure the unit test case still works. Thanks!

magibney reviewed (approved):
A few more comments. The only one that I think really needs a change is the `executorService.shutdown()`.

This basically LGTM, so I'll approve (informally conditional on fixing the `executorService.shutdown()`).

I do have some misgivings about the heavy use of `synchronized` methods and blocks -- I have a (admittedly vague) sense that it really should not be necessary. But maybe something to address before if/when upstreaming. For now, as employed per-individual-request, it should be perfectly fine.

magibney commented on solr/core/src/java/org/apache/solr/handler/component/TimeLimitingHttpShardHandlerFactory.java:
```diff
@@ -0,0 +1,180 @@
+        slowNodeTimeout,
+        dryRun,
+        slowNodeDetector,
+        (timedOutTasks) -> {
+          Counter counter = dryRun ? cancelledDryRunSlowNodeRequests : cancelledSlowNodeRequests;
+          counter.inc(timedOutTasks.size());
+        },
```
one other thought: it seems a little odd to me that we're passing `Set<Future<?>>` here, only to unconditionally invoke `timedOutTasks.size()`. Unless we anticipate using the set in some other way, could we just make this an `IntConsumer`? (and possibly also clarifying that this callback is strictly about metrics (since it seems that's the case with the current code).

Not a big deal either way.

patsonluk replied:
😁 I tried to be generic but `TimeLimitingHttpShardHandler` is a package private class anyway 🤷🏼 , so we really get to control what it does (no need to be generic).

Anyway, it's not a big deal. Passing the Set<Future<?>> (should be a small set for cancelled ones) for such a quick operation should not be costly. Let's keep it as is for now!

patsonluk committed:
Changes after code review
